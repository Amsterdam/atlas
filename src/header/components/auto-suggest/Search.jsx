import React from 'react'
import PropTypes from 'prop-types'
import styled from '@datapunt/asc-core'
import { SearchBar, SearchBarToggle, BackDrop, breakpoint } from '@datapunt/asc-ui'

// Actually we don't want the BackDrop to even render the Portal, but having multiple instances of components with BackDrops is stil something that can't be handled better so long the AutoSuggest component is still in this project
const StyledBackDrop = styled(BackDrop)`
  @media screen and ${breakpoint('max-width', 'tabletM')} {
    display: none;
  }
`

const StyledSearchBar = styled(SearchBar)`
  position: relative;

  @media screen and ${breakpoint('min-width', 'laptopM')} {
    width: 80%;
  }

  @media screen and ${breakpoint('min-width', 'tabletM')} {
    width: 80%;
  }
`

const Search = ({
  expanded,
  onBlur,
  searchBarProps,
  openSearchBarToggle,
  onOpenSearchBarToggle,
  inputProps,
  children,
}) => {
  const onOpenSearchToggle = open => {
    onOpenSearchBarToggle(open)
  }

  return (
    <React.Fragment>
      <StyledSearchBar
        showAt="tabletM"
        inputProps={inputProps}
        {...searchBarProps}
        aria-haspopup="true"
        aria-expanded={expanded}
        onMouseLeave={onBlur} // the onMouseLeave event is needed here as the MenuInline component has onMouseEnter/onMouseLeave events and we don't want a menu and the AutoSuggest to be opened at the same time
      >
        {expanded && (
          <StyledBackDrop
            onMouseEnter={onBlur} // The BackDrop is rendered as a Portal, therefore the onMouseLeave event isn't fired when entering it and we must call this event instead
            hideOverFlow={false}
          />
        )}
        {children}
      </StyledSearchBar>
      <SearchBarToggle
        hideAt="tabletM"
        onOpen={onOpenSearchToggle}
        open={openSearchBarToggle}
        inputProps={inputProps}
        searchBarProps={searchBarProps}
        aria-haspopup="true"
        aria-expanded={expanded}
        hasBackDrop
      />
    </React.Fragment>
  )
}

Search.propTypes = {
  expanded: PropTypes.bool.isRequired,
  searchBarProps: PropTypes.shape({}).isRequired,
  inputProps: PropTypes.shape({}).isRequired,
  onOpenSearchBarToggle: PropTypes.func.isRequired,
  openSearchBarToggle: PropTypes.bool.isRequired,
}

export default Search
