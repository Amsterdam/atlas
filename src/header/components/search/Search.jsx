import React from 'react';
import PropTypes from 'prop-types';
import AutoSuggestWrapper from '../../wrappers/auto-suggest/AutoSuggestWrapper';
import getSharedConfig from '../../../shared/services/shared-config/shared-config';
import autoSuggestService from '../../services/auto-suggest/auto-suggest-service';

import './_search.scss';

class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      query: props.searchQuery,
      suggestions: []
    };

    this.onTextInput = this.onTextInput.bind(this);
    this.onSuggestSelection = this.onSuggestSelection.bind(this);
    this.getSuggestions = this.getSuggestions.bind(this);
    this.onFormSubmit = this.onFormSubmit.bind(this);
    this.clearSuggestions = this.clearSuggestions.bind(this);
    if (window.opener && window.suggestionToLoadUri) {
      this.openDetailOnLoad();
    }
  }

  onTextInput(event) {
    const { onSearchInput, setSuggestions } = this.props;
    if (!event || event.target.value === '') {
      // clear
      onSearchInput();
      setSuggestions();
    } else {
      onSearchInput(event.target.value);
      // timeout to ensure the state has been set before the getSuggestions fn is called
      setTimeout(this.getSuggestions, 0, false);
    }
  }

  onSuggestSelection(suggestion, event) {
    const { fetchDetail, searchQuery } = this.props;
    event.preventDefault();
    event.stopPropagation();

    // eslint-disable-next-line no-underscore-dangle
    window._paq.push(['trackEvent', 'search', 'auto-suggest', searchQuery, suggestion._display]);

    if (event.ctrlKey || event.metaKey) {
      const newWindow = window.open(`${window.location.href}`, '_blank');
      // setting uri to the window, as window.postMessage does not work for some reason
      // (webpack overrides the data it seems)
      newWindow.window.suggestionToLoadUri = suggestion.uri;
    } else {
      fetchDetail(`${getSharedConfig().API_ROOT}${suggestion.uri}`);
      this.clearSuggestions();
    }
  }

  onFormSubmit(event) {
    event.preventDefault();
    event.stopPropagation();
    const { fetchDataSelection,
      fetchSearchResultsByQuery,
      isDatasetView,
      searchQuery,
      emptyFilters,
      numberOfSuggestions,
      activeSuggestionIndex
    } = this.props;

    // eslint-disable-next-line no-underscore-dangle
    window._paq.push(['trackSiteSearch', searchQuery, isDatasetView ? 'datasets' : 'data', numberOfSuggestions]);

    if (activeSuggestionIndex === -1) {
      // Load the search results
      emptyFilters();
      if (isDatasetView) {
        fetchDataSelection(searchQuery);
      } else {
        fetchSearchResultsByQuery(searchQuery);
      }
    }
    this.clearSuggestions();
  }

  getSuggestions() {
    const { searchQuery, setSuggestions, setActiveSuggestionIndex } = this.props;
    setActiveSuggestionIndex();

    this.setState({
      originalQuery: this.state.query
    });

    if (searchQuery.length > 1) {
      autoSuggestService.search(searchQuery).then((suggestions) => {
        if (suggestions && suggestions.query === searchQuery) {
          // Only load suggestions if they are still relevant.
          setSuggestions(suggestions.data, suggestions.count);
        }
      });
    } else {
      setSuggestions();
    }
  }

  clearSuggestions() {
    const { onSearchInput, setSuggestions, setActiveSuggestionIndex } = this.props;

    onSearchInput();
    setSuggestions();
    setActiveSuggestionIndex();
  }

  openDetailOnLoad() {
    const { fetchDetail } = this.props;
    const suggestionUri = window.suggestionToLoadUri;
    fetchDetail(`${getSharedConfig().API_ROOT}${suggestionUri}`);
    window.suggestionToLoadUri = null;
  }

  render() {
    const {
      suggestions,
      searchQuery,
      setActiveSuggestionIndex,
      activeSuggestionIndex,
      showSuggestions,
      setShowSuggestions,
      numberOfSuggestions,
      setActiveSuggestion,
      activeSuggestion
    } = this.props;

    return (
      <AutoSuggestWrapper
        placeHolder={'Zoek data op adres, postcode, kadastrale aanduiding, etc. Of datasets op trefwoord.'}
        classNames={'c-search-form__input js-search-input qa-search-form-input'}
        uniqueId={'global-search'}
        legendTitle={'Data zoeken'}
        onTextInput={this.onTextInput}
        suggestions={suggestions}
        numberOfSuggestions={numberOfSuggestions}
        query={searchQuery}
        onSuggestSelection={this.onSuggestSelection}
        setActiveSuggestionIndex={setActiveSuggestionIndex}
        setActiveSuggestion={setActiveSuggestion}
        activeSuggestionIndex={activeSuggestionIndex}
        activeSuggestion={activeSuggestion}
        showSuggestions={showSuggestions}
        setShowSuggestions={setShowSuggestions}
        onSubmit={this.onFormSubmit}
      />
    );
  }
}

Search.propTypes = {
  searchQuery: PropTypes.string,
  suggestions: PropTypes.arrayOf(PropTypes.object),
  numberOfSuggestions: PropTypes.number,
  onSearchInput: PropTypes.func.isRequired,
  setSuggestions: PropTypes.func.isRequired,
  fetchDetail: PropTypes.func.isRequired,
  fetchDataSelection: PropTypes.func.isRequired,
  fetchSearchResultsByQuery: PropTypes.func.isRequired,
  emptyFilters: PropTypes.func.isRequired,
  isDatasetView: PropTypes.bool,
  showSuggestions: PropTypes.bool,
  activeSuggestionIndex: PropTypes.number,
  setActiveSuggestionIndex: PropTypes.func.isRequired,
  setActiveSuggestion: PropTypes.func.isRequired,
  setShowSuggestions: PropTypes.func.isRequired,
  activeSuggestion: PropTypes.object //eslint-disable-line
};

Search.defaultProps = {
  searchQuery: '',
  suggestions: [],
  isDatasetView: false,
  numberOfSuggestions: 0,
  activeSuggestionIndex: -1,
  showSuggestions: false,
  activeSuggestion: {}
};


Search.contextTypes = {
  store: PropTypes.object.isRequired
};

export default Search;
