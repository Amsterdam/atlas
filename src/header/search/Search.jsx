import React from 'react';
import PropTypes from 'prop-types';
import AutoSuggest from '../../shared/components/autosuggest/AutoSuggest';
import getSharedConfig from '../../shared/services/shared-config/shared-config';
import ACTIONS from '../../shared/actions';
import autosuggestDataService from './autosuggest-service';

class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      query: props.query,
      suggestions: ''
    };
    this.onTextInput = this.onTextInput.bind(this);
    this.onSuggestSelection = this.onSuggestSelection.bind(this);
    this.getSuggestions = this.getSuggestions.bind(this);
  }

  onTextInput(event) {
    if (!event || event.target.value === '') {
      // clear
      this.setState({
        suggestions: [],
        query: ''
      });
    } else {
      this.setState({
        query: event.target.value
      });
      this.getSuggestions(event.target.value);
    }
  }

  onSuggestSelection(suggestion) {
    this.context.store.dispatch({
      type: ACTIONS.FETCH_DETAIL,
      payload: `${getSharedConfig().API_ROOT}${suggestion.uri}`
    });
  }

  getSuggestions(query) {
    // scope.activeSuggestionIndex = -1;
    // scope.originalQuery = scope.query;

    if (query.length > 1) {
      autosuggestDataService.search(query).then((suggestions) => {
        // Only load suggestions if they are still relevant.
        if (suggestions && suggestions.query === query) {
          this.setState({
            suggestions: suggestions.data
          });
        }
      });
    } else {
      this.setState({
        suggestions: [],
        numberOfSuggestions: 0
      });
    }
  }

  render() {
    // const { inputId } = this.props;

    return (
      <div id={'header-search'}>
        <form className="c-search-form" ng-submit="query.trim() && formSubmit($event)">
          <fieldset>
            <AutoSuggest
              placeHolder={'Zoek data op adres, postcode, kadastrale aanduiding, etc. Of datasets op trefwoord.'}
              classNames={'c-search-form__input js-search-input qa-search-form-input'}
              uniqueId={'global-search'}
              legendTitle={'Data zoeken'}
              onTextInput={this.onTextInput}
              suggestions={this.state.suggestions}
              query={this.state.query}
              onSuggestSelection={this.onSuggestSelection}
            />
            <button
              disabled={!this.state.query.trim()}
              className="c-search-form__submit qa-search-form-submit"
              type="submit"
              title="Zoeken"
            >
              <span className="u-sr-only">Zoeken</span>
            </button>
          </fieldset>
        </form>
      </div>
    );
  }
}

Search.propTypes = {
  query: PropTypes.string
};

Search.defaultProps = {
  query: ''
};


Search.contextTypes = {
  store: PropTypes.object.isRequired
};

export default Search;
