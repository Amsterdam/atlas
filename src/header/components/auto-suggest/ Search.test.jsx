import { ThemeProvider } from '@datapunt/asc-ui'
import { render } from '@testing-library/react'
import React from 'react'
import Search from './Search'

describe('Search', () => {
  const onOpenSearchBarToggleMock = jest.fn()
  const props = {
    expanded: false,
    searchBarProps: {},
    openSearchBarToggle: false,
    onOpenSearchBarToggle: onOpenSearchBarToggleMock,
    inputProps: {
      placeholder: 'Zoek',
    },
  }

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should not show the backdrop by default', () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <Search {...props} />
      </ThemeProvider>,
    )

    expect(getByTestId('backDrop')).toHaveStyleRule('display', 'none')
  })

  it('should show the backdrop when the component receives the right props', () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <Search {...{ ...props, expanded: true }} />
      </ThemeProvider>,
    )

    expect(getByTestId('backDrop')).toHaveStyleRule('display', 'initial')
  })
})
