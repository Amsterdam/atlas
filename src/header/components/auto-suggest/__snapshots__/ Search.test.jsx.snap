// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should shallow the searchbar and searchtoggle 1`] = `
.c1 {
  display: block;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0,0,0,0.7);
  z-index: 20;
}

.c0 {
  display: none;
}

@media screen and (max-width:767px) {
  .c0 {
    display: none;
  }
}

<Styled(BackDrop)
  expanded={false}
  hideOverFlow={false}
>
  <StyledComponent
    expanded={false}
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-hXRMBi",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "
  ",
            [Function],
            "

  @media screen and ",
            [Function],
            " {
    display: none;
  }
",
          ],
        },
        "displayName": "Styled(BackDrop)",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-hXRMBi",
        "target": [Function],
        "toString": [Function],
        "usesTheme": false,
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    hideOverFlow={false}
  >
    <BackDrop
      className="c0"
      expanded={false}
      hideOverFlow={false}
    >
      <Portal
        hideOverFlow={false}
      >
        <Portal
          containerInfo={
            .c0 {
  display: block;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: rgba(0,0,0,0.7);
  z-index: 20;
}

.c1 {
  display: none;
}

@media screen and (max-width:767px) {
  .c1 {
    display: none;
  }
}

<div
              style="position: absolute;"
            >
              <div
                class="c0 c1"
              />
            </div>
          }
        >
          <styled.div
            className="c0"
            expanded={false}
          >
            <StyledComponent
              className="c0"
              expanded={false}
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-htpNat",
                    "isStatic": false,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
  display: block;
  ",
                      "position: fixed;",
                      "top: 0;",
                      "right: 0;",
                      "bottom: 0;",
                      "left: 0;",
                      ";
  background-color: ",
                      [Function],
                      ";
  z-index: ",
                      "20",
                      ";
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-htpNat",
                  "target": "div",
                  "toString": [Function],
                  "usesTheme": false,
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="c1 c0"
              />
            </StyledComponent>
          </styled.div>
        </Portal>
      </Portal>
    </BackDrop>
  </StyledComponent>
</Styled(BackDrop)>
`;
